query: Allowing duplicated interface name with different interface types

cathay4t opened the pull request:
As only OVS internal interface has kernel nic entry, it is allowed
for a OVS bridge hold the same name as OVS internal interface.

This patch change the upper level of nmstate to allow duplicated
interface name and only enforce unique interface name within the same
interface type.

When different plugins providing information regarding the same
interface, it is required they also provide the interface type.

If any unknown interface has single interface with valid name and type,
information will be merged also.

The nispor plugin has changed to support OVS internal interface.
The OVS db plugin has changed to include interface type also.

cathay4t commented:
~~I am still working on test cases.~~

ffmancera commented on libnmstate/nm/ovs.py:
```diff
@@ -184,46 +185,50 @@ def _get_patch_setting(act_con):
 
-def get_port(nm_device):
-    return nm_device.get_slaves()
+def _get_lag_nmstate_port_info(nm_dev_ovs_port):
+    port_info = {}
```
Does this make sense?

ffmancera commented on libnmstate/nm/ovs.py:
```diff
@@ -184,46 +185,50 @@ def _get_patch_setting(act_con):
+def _get_lag_nmstate_port_info(nm_dev_ovs_port):
+    port_info = {}
 
+    OVS_LAG = OB.Port.LinkAggregation
```
Is a function modifying a constant?

cathay4t replied:
Just create alias for the long constant.

ffmancera commented on libnmstate/nm/ovs.py:
```diff
@@ -125,36 +124,38 @@ def create_patch_setting(patch_state):
+        if len(nm_dev_ovs_ifaces) == 1:
+            port_info[OB.Port.NAME] = nm_dev_ovs_ifaces[0].get_iface()
+        else:
+            port_info = _get_lag_nmstate_port_info(nm_dev_ovs_port)
```
How this works? I think this functions is returning `None`.

cathay4t replied:
There is a `return ports_info` below.

EdDev commented on tests/integration/ovs_test.py:
```diff
@@ -630,3 +630,46 @@ def test_remove_all_ovs_ports(bridge_with_ports):
+        cmdlib.exec_cmd(
+            ["nmcli", "connection", "delete"] + con_to_delete, check=True,
+        )
+
```
I found it raceful without a 1sec sleep. It is fine to keep it without but keep an eye on this in case the assert fails.

cathay4t replied:
I will create the support of creating ovs interface with the same name of bridge via nmstate. This is planned feature for 0.4 branch.

Once that include, we will not use nmcli any more for this test case.

EdDev commented on tests/integration/ovs_test.py:
```diff
@@ -630,3 +630,46 @@ def test_remove_all_ovs_ports(bridge_with_ports):
+def test_ovs_internal_using_the_same_name_as_bridge(
+    ovs_bridge_with_internal_interface_and_identical_names,
+):
+    libnmstate.show()
```
I think we should check that it detects 2 interfaces of the expected name (returned by the fixture).

EdDev commented on libnmstate/nm/ovs.py:
```diff
@@ -125,36 +124,38 @@ def create_patch_setting(patch_state):
 
+    if ports_info or options:
+        ovs_info[OB.PORT_SUBTREE] = ports_info
+        ovs_info[OB.OPTIONS_SUBTREE] = options
```
nit: Instead of line 133, you could do it here directly:
```
iface_info[OB.CONFIG_SUBTREE] = {
    OB.PORT_SUBTREE: ports_info,
    OB.OPTIONS_SUBTREE: options,
}
```

It is also interesting why we need to specify both if only one exists, but that is the current code as well.

EdDev commented on libnmstate/nm/ovs.py:
```diff
@@ -184,46 +185,50 @@ def _get_patch_setting(act_con):
+                {OVS_LAG.Port.NAME: nm_dev_ovs_iface.get_iface()}
+                for nm_dev_ovs_iface in nm_dev_ovs_port.get_slaves()
+            ],
+            key=itemgetter(OB.Port.LinkAggregation.Port.NAME),
```
nit: `OVS_LAG.Port.NAME`

EdDev commented on libnmstate/nm/ovs.py:
```diff
@@ -184,46 +185,50 @@ def _get_patch_setting(act_con):
+    mode = _get_lag_mode(nm_dev_ovs_port)
+    if mode:
+        port_info[OB.Port.LINK_AGGREGATION_SUBTREE][
+            OB.Port.LinkAggregation.MODE
```
nit: `OVS_LAG.MODE`

EdDev commented on libnmstate/nm/ovs.py:
```diff
@@ -184,46 +185,50 @@ def _get_patch_setting(act_con):
-    ports_info.sort(key=itemgetter(OB.Port.NAME))
-    return ports_info
+    port_info[OB.Port.NAME] = nm_dev_ovs_port.get_iface()
+    port_info[OB.Port.LINK_AGGREGATION_SUBTREE] = {
```
Optionally, you do not have to use `port_info` at all and build it at the end.
You could use `lag = {....}` and at the very and just do `return {OB.Port.NAME: ... , OB.Port.LINK...: lag}`
It may just save a few ugly line breaks below.

EdDev commented on libnmstate/nm/ovs.py:
```diff
@@ -232,38 +237,26 @@ def _get_bridge_port_info(context, port_profile, devices_info):
+    if nm_ac:
+        nm_profile = nm_ac.props.connection
+        if nm_profile:
+            nm_setting = nm_profile.get_setting(NM.SettingOvsPort)
```
nit: You could directly return here without using a variable.

EdDev commented on libnmstate/nispor/plugin.py:
```diff
@@ -82,6 +83,8 @@ def get_interfaces(self):
                 )
             elif iface_type == "Vrf":
                 ifaces.append(NisporPluginVrfIface(np_iface).to_dict())
+            elif iface_type == {"Other": "openvswitch"}:
```
Why is this a dict? (all others are simple strings)

cathay4t replied:
This is because nispor place unknown data as {"Other": xxxx}. But it has been supported by nispor 0.6.1.

EdDev commented on libnmstate/nm/profile.py:
```diff
@@ -619,7 +619,10 @@ def get_all_applied_configs(context):
-                action = f"Retrieve applied config: {iface_name}"
+                iface_type_str = nm_dev.get_type_description()
+                action = (
+                    f"Retrieve applied config: {iface_type_str} {iface_name}"
```
Nice, you use now an identifier which uses both the name and type.
As it is not trivial, consider adding in the commit message a note about this or add here a note inline.
This is specific to nm, so it may even fit a separate commit.
Whatever fits you best.

EdDev commented on libnmstate/nm/profile_state.py:
```diff
@@ -280,6 +280,7 @@ def get_applied_config_callback(nm_dev, result, user_data):
     context.finish_async(action)
     try:
         remote_conn, _ = nm_dev.get_applied_connection_finish(result)
+        # TODO: We should both interface name and type as key below.
```
How come this is not a must now?

cathay4t replied:
We don't use applied config for OVS interfaces yet. When we does, we will need it.

cathay4t replied:
Just keep this PR small.

EdDev commented on libnmstate/nmstate.py:
```diff
@@ -256,3 +287,19 @@ def _get_routes_from_plugins(plugins):
+    """
+    iface_types = []
+    for iface_info in iface_infos:
+        iface_type = iface_info.get(Interface.TYPE, InterfaceType.UNKNOWN)
+        if (
+            iface_type != InterfaceType.UNKNOWN
```
maybe keep the default None, it will be simpler.

EdDev replied:
Was this missed?

cathay4t replied:
I leave it as it was as I have to handle two cases:
 * Type not defined.
 * Type been defined as `UNKNOWN` explicitly.

Using default None cannot handle the second case simply.

EdDev commented on libnmstate/nmstate.py:
```diff
@@ -166,17 +167,47 @@ def _get_interface_info_from_plugins(plugins):
+
+    # For any unknown interface, we merge it when there is only one interface
+    # with the same name and valid type
+    to_be_removed_index = []
```
This section is a bit complicated and merits its own function and a reasoning.

In addition:
- Why is this needed? What does it solve etc...
  (I think it is because the UNKNOWN ones are reported by the ovsdb plugin for system interfaces, so the idea is to overwrite an UNKNOWN type with a known type)
  Anyway, it requires documentation as this is a generic rule and it is not trivial.
- Can it be unit tested, covering all the flow paths?

cathay4t replied:
This is need for ovsdb plugin to include extra information for system interface like eth1. It's hard not waste of time for ovsdb plugin to detect the interface type of a ovs system interface.

The third party plugin might don't want to detect the interface type neither. But doing above complex guess in nmstate top level, we easy the plugin coder's life.

cathay4t replied:
I will add the unit test cases for this.

EdDev commented on libnmstate/nmstate.py:
```diff
@@ -166,17 +167,47 @@ def _get_interface_info_from_plugins(plugins):
+            iface_types = _get_iface_types_by_name(
+                all_ifaces.values(), iface_name
+            )
+            if len(iface_types) == 1:
```
What if there are 2 such types or even more? It may be safer to use `> 0`, no?

cathay4t replied:
As it state in the comment. We only merge unknown interface data to valid interface when there is only __single__ valid interface available.

EdDev commented on libnmstate/plugins/nmstate_plugin_ovsdb.py:
```diff
@@ -147,12 +148,30 @@ def plugin_capabilities(self):
+            elif row.type == "system":
+                # Let other plugin decide the interface type
+                iface_type = InterfaceType.UNKNOWN
+            else:
```
Where is the `patch` interface type fits here?

cathay4t replied:
Good catch. Will shod `patch` in `ovs-interface` also.

EdDev commented on tests/integration/ovs_test.py:
```diff
@@ -630,3 +630,46 @@ def test_remove_all_ovs_ports(bridge_with_ports):
+        yield device_name
+    finally:
+        _, con_names, _ = cmdlib.exec_cmd(
+            ["nmcli", "-f", "NAME", "connection"], check=True,
```
Better be consistent and keep using `split`. (or the other way around)

EdDev commented on tests/integration/ovs_test.py:
```diff
@@ -630,3 +630,46 @@ def test_remove_all_ovs_ports(bridge_with_ports):
+            if device_name in con_name
+        ]
+        cmdlib.exec_cmd(
+            ["nmcli", "connection", "delete"] + con_to_delete, check=True,
```
Same here.

cathay4t committed:
nispor vxlan: Sync with nispor 0.6.1

The nispor has deprecated the incorrect `dst_port_min` property.
We should use `dst_port` instead.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
nm ovs: Support OVS port using the same name of OVS interface

Previously, we are using interface name to determined the ports of OVS
bridge. This become a problem when the OVS port profile is sharing the
name with OVS interface.

This patch changed it to use `get_slaves()` of `NM.DeviceOvsBridge` and
`NM.DeviceOvsPort` to clarify this indistinct.

Also removed unused functions.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
nm profile: Use iface type and name as action ID

Use iface type and name as action for retrieving applied configure.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
New patch uploaded. Please review again.

cathay4t committed:
query: Allowing duplicated interface name with different interface types

As only OVS internal interface has kernel nic entry, it is allowed
for a OVS bridge hold the same name as OVS internal interface.

This patch change the upper level of nmstate to allow duplicated
interface name and only enforce unique interface name within the same
interface type.

When different plugins providing information regarding the same
interface, it is required they also provide the interface type.

If any unknown interface has single interface with valid name and type,
information will be merged also.

The nispor plugin has changed to support OVS internal interface.
The OVS db plugin has changed to include interface type also.

Integration and unit test cases included.

Signed-off-by: Gris Ge <fge@redhat.com>

EdDev reviewed (approved):
Thanks, looks great.
There is one point missed in there, but it is a nit. Up to you if you want to handle it.
